source/opt/local_single_store_elim_pass.cpp

```diff
@@ -0,0 +1,494 @@
+    // Don't delete labels
+    if (di->opcode() == SpvOpLabel) {
+      deadInsts.pop();
+      continue;
+    }
+    // Remember operands
+    std::queue<uint32_t> ids;
+    di->ForEachInId([&ids](uint32_t* iid) {
+      ids.push(*iid);
+    });
+    uint32_t varId = 0;
+    // Remember variable if dead load
+    if (di->opcode() == SpvOpLoad)
+      (void) GetPtr(di, &varId);
+    def_use_mgr_->KillInst(di);
+    // For all operands with no remaining uses, add their instruction
+    // to the dead instruction queue.
+    while (!ids.empty()) {
+      uint32_t id = ids.front();
+      analysis::UseList* uses = def_use_mgr_->GetUses(id);
+      if (uses == nullptr)
+        deadInsts.push(def_use_mgr_->GetDef(id));
+      ids.pop();
+    }
+    // if a load was deleted and it was the variable's
+    // last load, add all its stores to dead queue
+    if (varId != 0 && !IsLiveVar(varId)) 
+      AddStores(varId, &deadInsts);
+    deadInsts.pop();
+  }
+}
+
+bool LocalSingleStoreElimPass::SingleStoreDCE() {
+  bool modified = false;
+  for (auto v : ssa_var2store_) {
+    // check that it hasn't already been DCE'd
+    if (v.second->opcode() != SpvOpStore)
+      continue;
+    if (non_ssa_vars_.find(v.first) != non_ssa_vars_.end())
+      continue;
+    if (!IsLiveStore(v.second)) {
+      DCEInst(v.second);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+bool LocalSingleStoreElimPass::HasUnsupportedInst(ir::Function* func) {
+  // Currently this pass only supports optimization of store and load.
+  // The presence of other memory operations as well as function calls and
+  // non-access-chain pointer operations is not currently supported and
+  // will cause the function to return unmodified.
+  // TODO(): Handle more memory operations, pointer operations, function
+  // calls.
+  for (auto& blk : *func)
+    for (auto& inst : blk)
+      switch (inst.opcode()) {
+      case SpvOpAtomicLoad:
+      case SpvOpAtomicStore:
+      case SpvOpAtomicExchange:
+      case SpvOpAtomicCompareExchange:
+      case SpvOpAtomicCompareExchangeWeak:
+      case SpvOpAtomicIIncrement:
+      case SpvOpAtomicIDecrement:
+      case SpvOpAtomicIAdd:
+      case SpvOpAtomicISub:
+      case SpvOpAtomicSMin:
+      case SpvOpAtomicUMin:
+      case SpvOpAtomicSMax:
+      case SpvOpAtomicUMax:
+      case SpvOpAtomicAnd:
+      case SpvOpAtomicOr:
+      case SpvOpAtomicXor:
+      case SpvOpLifetimeStart:
+      case SpvOpLifetimeStop:
+      case SpvOpCopyMemory:
+      case SpvOpFunctionCall:
+        return true;
+      // This is a pre-SSA pass, so Conventional SSA form including
+      // OpCopyObject is not supported.
+      case SpvOpCopyObject:
+        return true;
+      // Frexp and Modf extended ops reference memory and are unsupported
+      case SpvOpExtInst: {
+        uint32_t op = inst.GetSingleWordInOperand(kSpvExtInstInstruction);
+        if (op == GLSLstd450Frexp || op == GLSLstd450Modf)
```

Comment:
Hmmm.  How do you know this extended instruction is from GLSL.std.450?

There are similar bad cases in OpenCL extended instructions, e.g. frexp, fract, lgamma_r, and a few dozen more(!)

Overall, the strategy of looking for a few bad instructions is both too weak and too strong. Too weak because it doesn't cover new cases you can't anticipate, like new extended instruction sets.  Too strong because an AtomicXor on a variable in StorageBuffer storage class shouldn't affect the ability to eliminate loads and stores in Function storage class.

I think it's best to go with what I suggested earlier:  In SingleStorageAnalyze, scan instructions other than Load (and Store), and insert any id arguments into non_ssa_vars_.  That should fail safely:  newly invented instructions that use a pointer-to-variable will cause the algorithm to gracefully avoid the optimization.

I realize this reverses my previous approval to a pre-pass to bail on bad cases. Now that I see the code I see the problem with it.